### ISSUE TITLE
Cart rule restrictions wrongly saved when 40+ cart rules

### ISSUE THREAD

**Issue by Flower7C3** [PrestaShop/PrestaShop#17603]
**Describe the bug**
Unable to save selected cart rule restrictions.

**To Reproduce**
**PS**: You should have more than 40 cart rules to reproduce the issue

1. Go to BO, Catalogue, Cart Rules
2. Add new rule
3. Check restrictions and move 1 restriction to left
4. Save
5. Edit this rule
6. Multiple restrictions on left

**Screenshots**
Remove 1 restriction
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/3767861/74240511-6fc53c80-4cda-11ea-95d1-66c0b829f815.png">

I have saved 190 restrictions in DB
<img width="255" alt="image" src="https://user-images.githubusercontent.com/3767861/74240913-2f19f300-4cdb-11ea-897a-e000f1249df0.png">

So my rule should has 188 restrictions saved in database, but has only 39
<img width="192" alt="image" src="https://user-images.githubusercontent.com/3767861/74240579-984d3680-4cda-11ea-8a49-c199f456387a.png">

And visible in BO
<img width="805" alt="image" src="https://user-images.githubusercontent.com/3767861/74240615-a7cc7f80-4cda-11ea-9185-7f732fe8d6a4.png">

Problem is probably in select, which is loading asynchronously elements when I scroll to the bottom
<img width="321" alt="image" src="https://user-images.githubusercontent.com/3767861/74241079-9172f380-4cdb-11ea-94d7-54dc51e1e418.png">

**Additional information**
PrestaShop version: 1.7.7.0
PHP version: 7.2

**Comment by khouloudbelguith** [comment]
Hi @Flower7C3,

I did not manage to reproduce the issue with PS1.7.6.3.
I attached a screen record
https://drive.google.com/file/d/1G8zmvyy0hsvrpbROSCf20PcdMmyim0Ne/view
Thanks to check & feedback.

**Comment by Flower7C3** [comment]
I am using PS 1.7.6.2. During the weekend I am planning to update to 1.7.6.3 so maybe the problem will disappear or it is caused by some module.

I will back later with info.

**Comment by khouloudbelguith** [comment]
@Flower7C3, thanks!
Waiting for your feedback.

**Comment by Flower7C3** [comment]
Hi @khouloudbelguith, after upgrading to PS 1.7.6.3 problem still exists on my instance.

Could You do recheck, but with more cart rules? I think that problem is when there are more than 40 cart rules in database. So not all items are in right select during page load. And they are being loaded when You scroll down on the list (http://s000.tinyupload.com/index.php?file_id=32136249853765246079).

**Comment by khouloudbelguith** [comment]
@Flower7C3, 

> I think that problem is when there are more than 40 cart rules in database

No, I tried with PS1.7.6.3 & I have 41 cart rules & it is OK.
I attached a screen record
[cart_rule.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4213389/cart_rule.webm.zip)
We need to retrieve the PHP error log and the debug mode report in order to find out what's wrong.
Don't you know how to get this information? Please read the following article:
http://build.prestashop.com/howtos/misc/how-to-create-bug-report/

Thanks!

**Comment by Flower7C3** [comment]
But when You scroll down to **test** cart rule it probably cause AJAX request for more rules. Please do not scroll down.

**Comment by khouloudbelguith** [comment]
@Flower7C3, thanks for your feedback.
I manage to reproduce the issue with PS1.7.6.3 & PS1.7.5.2
If you have more than 40 cart rules, All the `cart rules > 40` will be displayed in the `Uncombinable cart rules`.
https://drive.google.com/file/d/1vpvNtWJzmTPIgVEZmaHWGX8zxXCEQt2X/view
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by khouloudbelguith** [comment]
Hi!

Same issue with 1770beta2build1

**Comment by csoler-webimpacto** [comment]
Hi, I have reproduced this bug un Prestashop 8.0.4. 
Any solution, please?

**Comment by medotek** [comment]
Hi, the following listed issues #31201 #15526 have the same problem.

Anyway, i'm still confronted to this issue (PS 1.7.8.7). I've managed to fixed that but still inappropriate to be in a pull request. 
I think this has to be reworked for the next versions.

It's about options that fully disappear when there is a research on input. 
To fixed that, i display the cart rules without limit (because the issue doesn't necessarily appear when there is more than 40 cart rules). 
You can test with 3 cart rules, 2 on the uncompatible and the other on the compatible. Then search one cart rule among the uncompatible ones (in order to remove them from the options).
I changed the behaviour of the filters in order to hide the options instead of removing them.

```js
// admin/themes/default/template/controllers/cart_rules/form.js
function combinable_filter(id_rule, search, type) {
  //  Prevent from options to be gone, creating bugs/disabling restrictions ...
  const options = $(id_rule).children('option')

  if (options.length) {
    options.filter(function() {
      $(this).toggle($(this).text().toLowerCase().indexOf(search.toLowerCase()) > -1)
    });
  }
}
```

```php
// controllers/admin/AdminCartRulesController.php method ajaxProcessLoadCartRules
$limit = null;
$count = $id_cart_rule = 0;
// Comment this condition
//        if (Tools::getIsset('limit')) {
//            $limit = Tools::getValue('limit');
//        }
```

I hope we can find a real solution for this issue.

**Comment by tarekfellah** [comment]
Is there any fix for this issue?
@medotek i tried your solution, but it didn't fix the issue, i have the filter working but on click on save (or save and stay) i got blank page.

**Comment by medotek** [comment]
Hi @tarekfellah, did you catch the error ? from logs or with dev mode ? How many cartRules do you have in total ?

**Comment by tarekfellah** [comment]
@medotek my issues is a bit different, i haven't an issue on the filter but i when i moved a restriction to the right and when i clicked on the save (also save and stay) i got a blank page, even with the debug mod enabled, no console log error no ajax call, it's a strange issue.
I tried to add a vardump in postProcess function 
```
if (Tools::isSubmit('submitAddcart_rule') || Tools::isSubmit('submitAddcart_ruleAndStay')) {
        	var_dump('php here');exit;
```
I got only one time that i get the dump executed for many attempts

**Comment by medotek** [comment]
@tarekfellah try to check your apache/nginx error log

**Comment by tarekfellah** [comment]
@medotek i tried to check the var/logs/dev

**Comment by medotek** [comment]
@tarekfellah you should check deeper, some errors might be logged by your virtual host (apache/nginx) that's running on your local project or on a server. They are generally located in (the following absolutes paths) /var/log/apache2 or /var/log/nginx, depending on what vhost you're using.

**Comment by Daaaaad** [comment]
@Flower7C3 @csoler-webimpacto I have the same problem on a PS 1.7.8.7. Do you managed to solve the problem on your side? Thx, David.

**Comment by Flower7C3** [comment]
@Daaaaad sorry, I did not write any solution for this.

But:
1. in **Limitations** section
2. when **Compatibility with other cart policies** checkbox is checked
3. the **Connecting cart rules** select will appear
Then You can scroll down on that list, so all rules will by loaded with AJAX.

![image](https://github.com/PrestaShop/PrestaShop/assets/3767861/5ebf86d8-4dbe-4a5c-9985-52a045449a41)

**Comment by Daaaaad** [comment]
@Flower7C3 So you mean each time we want to save cart rules we have to scroll all the way down to load all the rules and then make the modifications and save? Doing so, the bug will no longer occur, right? Regards, David.

**Comment by Flower7C3** [comment]
@Daaaaad yes, that solution I have posted 4 years ago to my client in Trello.

**Comment by Daaaaad** [comment]
Ok thank you @Flower7C3 🙏 I'll pass on the information to my customer... but we have 2000+ rules so the scroll will take a very loooooooong time 😴

@khouloudbelguith What's the status of this bug? Is it fixed on current versions?

Regards,

David.

**Comment by Hlavtox** [comment]
For more information, see also the related issue - https://github.com/PrestaShop/PrestaShop/issues/40093.

**Comment by Codencode** [comment]
Fixed by: https://github.com/PrestaShop/PrestaShop/pull/40099
